Add changelog for release v1.2.0-rc1 (#3327)
kubernetes-sigs/gateway-api: CHANGELOG/1.2-CHANGELOG.md

> # v1.2 Changelog

Reviewer: Mind making the corresponding version bumps throughout so this PR can be the one that triggers the release? (x-ref https://github.com/kubernetes-sigs/gateway-api/pull/3062/files)

Author: The RELEASE.md says that usually we get this in first though?

## Table of Contents

- [v1.2.0-rc1](#v120-rc1)

# v1.2.0-rc1

We expect that this release candidate is quite close to the final v1.2.0
release. However, subsequent breaking API changes are still possible.

This release candidate is suitable for implementors, but we do not recommend
shipping products based on a release candidate API due to the possibility of
incompatible changes prior to the final release. The following represents the
changes since v1.1.0:

> ## Release Cycle changes

Reviewer: Non-blocking nit, feel free to ignore: I feel like this could probably be further down the list as most readers will probably be more interested in the new user-facing features.

Author: I actually put this first on purpose so that folks who only look for features will end up looking at it more, so I'd rather leave it.

> In this release, we moved to using a phased release cycle modeled on the
> [upstream Kubernetes release cycle](https://kubernetes.io/releases/release/).

Reviewer: Maybe something like this:
```suggestion
This was our first release using a new [release cycle](https://gateway-api.sigs.k8s.io/contributing/release-cycle/) that is meant to make Gateway API releases more frequent and predictable.
```

  already is; we want to ensure that features make their way to Standard or
  are removed.
* **GEP Iteration and Review**: 5-7 weeks, where we work to update GEPs, and
  solidify and meet graduation criteria for in-scope GEPs.
* **API Refinement and Documentation**: 3-5 weeks, where we translate proposed
  changes from GEPs into changes to both the API specification and
  documentation.
* **SIG-Network Review and Release Candidates**: 2-4 weeks, where we perform
  our required upstream reviews, get any further required changes in, and
  release our release candidates so that implementations can get started
  with work on the new features asap.

For all the detail about this, please see the [Release Cycle docs](https://gateway-api.sigs.k8s.io/contributing/release-cycle/).

Relevant PRs:
* Adding Release Cycle Documentation by @robscott in [#3063](https://github.com/kubernetes-sigs/gateway-api/pull/3063)

## Standard Channel Additions and Changes

The Standard channel is Gateway API's set of maximally-stable install files.
Only features with the best testing and support are added to the standard
channel. This channel should be considered GA or stable, and future changes

> will only be compatible changes.

Reviewer: ```suggestion
will be fully backwards compatible.
```

> ### Infrastructure Labels and Annotations moved to Standard 🎉

Reviewer: Non-blocking nit, feel free to ignore: Since this is already in a "Graduating to standard" section, we probably don't need that repeated in each subheading.
```suggestion
### Infrastructure Labels and Annotations
```

[GEP-1867](https://gateway-api.sigs.k8s.io/geps/gep-1867/) added an
`infrastructure` stanza to Gateway objects that is intended to carry
infrastructure configuration specific to that Gateway object.

[GEP-1762](https://gateway-api.sigs.k8s.io/geps/gep-1762/) adds a section
for `labels` and `annotations` to this stanza that specifies labels and

> annotations to be annotated to all Kubernetes resources created to fulfill

Reviewer: Ideally this can also apply beyond Kubernetes
```suggestion
annotations to be annotated to all resources created to fulfill
```

that Gateway request.

This feature can be used to affect the labels and annotations created on

> Loadbalancer Services by in-cluster implementations to fulfill the Gateway
> contract.

Reviewer: ```suggestion
LoadBalancer Services by in-cluster implementations to fulfill the Gateway
contract or by Cloud Load Balancing resources created by Cloud Providers.
```

This feature has graduated to Standard and is now considered GA or Stable.

This feature's name for conformance tests and GatewayClass status reporting
is `GatewayInfrastructurePropagation`.

This feature's status is **Extended**, meaning that it is _optional_ for

> implementations to support. Please see the implementation's
> `gatewayclass.status.supportedFeatures` stanza or the 

Reviewer: ```suggestion
implementations to support. If you're using Experimental Channel, you can refer to the 
`supportedFeatures` field in the `status` of any GatewayClass.
```

> [implementation feature list](https://gateway-api.sigs.k8s.io/implementations/v1.2/)

Reviewer: Recommend leaving this off since it usually takes us at least one month to populate the page.

[GEP 1742](https://gateway-api.sigs.k8s.io/geps/gep-1742/).

The relevant feature names this for conformance tests and GatewayClass status
reporting are:
* `HTTPRouteRequestTimeout` for `httproute.spec.rules[].timeouts.request`
* `HTTPRouteBackendTimeout` for `httproute.spec.rules[].timeouts.backendRequest`

This feature's status is **Extended**, meaning that it is _optional_ for
implementations to support. Please see the implementation's
`gatewayclass.status.supportedFeatures` stanza or the 
[implementation feature list](https://gateway-api.sigs.k8s.io/implementations/v1.2/)
once the final release is out to see if your implementation supports it.

Relevant PRs:
* Graduate GEP-2257 and GEP-1742 to standard by @kflynn in [#3210](https://github.com/kubernetes-sigs/gateway-api/pull/3210)
* GEP2257 Duration parsing by @xtineskim in [#3271](https://github.com/kubernetes-sigs/gateway-api/pull/3271)

### BackendProtocol Support moved to Standard 🎉

The previous coordinated work across both Gateway API and upstream
Kubernetes which defined 3 new values for the AppProtocol field on
Service Ports has graduated to Standard and is now considered GA or Stable.

> The values are:
> * `kubernetes.io/h2c` - HTTP/2 over cleartext as described in

Reviewer: Nit: A lot of our docs gen requires an empty line before lists. It may not be necessary for changelogs, but just in case:
```suggestion
The values are:

* `kubernetes.io/h2c` - HTTP/2 over cleartext as described in
```

  [RFC7540](https://www.rfc-editor.org/rfc/rfc7540)
* `kubernetes.io/ws` - WebSocket over cleartext as described in
  [RFC6445](https://www.rfc-editor.org/rfc/rfc6455)
* `kubernetes.io/wss` - WebSocket over TLS as described in
  [RFC6455](https://www.rfc-editor.org/rfc/rfc6455)

These can now be used with Gateway API to describe the protocol to use for
connections to Kubernetes Services. For more information, refer to [GEP
1911](https://gateway-api.sigs.k8s.io/geps/gep-1911/).

The relevant feature names this for conformance tests and GatewayClass status

> reporting are:
> * `HTTPRouteBackendProtocolH2C` for H2C support, when

Reviewer: ```suggestion
reporting are:

* `HTTPRouteBackendProtocolH2C` for H2C support, when
```

  `service.spec.ports[].appProtocol` is `kubernetes.io/h2c`.
* `HTTPRouteBackendProtocolWebSocket` for Websocket support, when
  `service.spec.ports[].appProtocol` is `kubernetes.io/ws` or `kubernetes.io/wss`.

This feature's status is **Extended**, meaning that it is _optional_ for
implementations to support. Please see the implementation's
`gatewayclass.status.supportedFeatures` stanza or the 
[implementation feature list](https://gateway-api.sigs.k8s.io/implementations/v1.2/)
once the final release is out to see if your implementation supports it.

> Relevant PRs:
> * Backend Protocol Selection is no longer experimental and is now part of HTTPRoute Extended Conformance by @dprotaso in [#3108](https://github.com/kubernetes-sigs/gateway-api/pull/3108)

Reviewer: ```suggestion
Relevant PRs:

* Backend Protocol Selection is no longer experimental and is now part of HTTPRoute Extended Conformance by @dprotaso in [#3108](https://github.com/kubernetes-sigs/gateway-api/pull/3108)
```

### Other Standard channel changes

> * Increased the limit on HTTPRoute matches from 8 to 64 by @howardjohn in [#3205](https://github.com/kubernetes-sigs/gateway-api/pull/3205)

Reviewer: ```suggestion
* Increased the limit on HTTPRoute matches from 8 to 64 by @howardjohn in [#3205](https://github.com/kubernetes-sigs/gateway-api/pull/3205). Note that the total number of matches allowed across all Rules within a Route is still 128.
```

> * SupportedFeatures field in Gateway Class status are now presented as a list of objects by @LiorLieberman in [#3200](https://github.com/kubernetes-sigs/gateway-api/pull/3200)

Reviewer: I think this is still experimental only

This name, if present, may be used in `status` and logging to indicate which route
rule is being referenced in messages in a more readable way than an array index.

This feature has graduated to Experimental amd should now be used for
testing and verification purposes only. Experimental features may be
changed or removed in a future version.

This feature does not currently have a feature name defined.

This feature's status is **Extended**, meaning that it is _optional_ for
implementations to support.

As there is no feature name or conformance tests available for this feature
as yet, please see your implementation's documentation to see if it is
supported.

Relevant PRs:
* Implement of GEP-995 by @howardjohn in [#2985](https://github.com/kubernetes-sigs/gateway-api/pull/2985)
* Docs: route rule name by @guicassolato in [3299](https://github.com/kubernetes-sigs/gateway-api/pull/3299)

## Other specification changes

* Fix v1beta1 HTTPHeaderName alias by @spencerhance in [#3127](https://github.com/kubernetes-sigs/gateway-api/pull/3127)
* Document the route kinds compatible for each protocolType by @abs007 in [#2263](https://github.com/kubernetes-sigs/gateway-api/pull/2263)

> * Move GEP-995 to Implementable by @guicassolato in [#3166](https://github.com/kubernetes-sigs/gateway-api/pull/3166)

Reviewer: I think these kinds of GEP changes don't really belong in the changelog

* Add policy label to BackendLBPolicy to make it recognizable in gwctl by @gauravkghildiyal in [#3195](https://github.com/kubernetes-sigs/gateway-api/pull/3195)

> * remove unnecessary S in ProtocolType validation by @spacewander in [#3185](https://github.com/kubernetes-sigs/gateway-api/pull/3185)

Reviewer: I think this actually belongs in standard channel

* Remove extra spacing in the API Specification aliases by @nuno-faria in [#3218](https://github.com/kubernetes-sigs/gateway-api/pull/3218)
* fix: allow the header based session persistence not to have AbsoluteTimeout by @sanposhiho in [#3215](https://github.com/kubernetes-sigs/gateway-api/pull/3215)
* Update GRPCRoute GEP to standard by @xtineskim in [#3227](https://github.com/kubernetes-sigs/gateway-api/pull/3227)
* The generated conformance report API version is set to v1 instead of v1alpha1 by @mlavacca in [#3153](https://github.com/kubernetes-sigs/gateway-api/pull/3153)
* Use GRPCHeaderMatchType in GRPCHeaderMatch structure by @kevinpollet in [#3252](https://github.com/kubernetes-sigs/gateway-api/pull/3252)
* fix: inconsistent default reason for GatewayClass status (#3225) by @snorwin in [#3257](https://github.com/kubernetes-sigs/gateway-api/pull/3257)
* Fixing Mesh version references in docs by @robscott in [#3073](https://github.com/kubernetes-sigs/gateway-api/pull/3073)
* fix: Add BackendLBPolicy to experimental kustomization by @krouma in [#3083](https://github.com/kubernetes-sigs/gateway-api/pull/3083)

> ## Leadership changes

Reviewer: This is a really great touch, thanks for thinking of this!

In this release timeframe, Gateway API has been working on building our contributor
pool and promoting more contributors up our contributor ladder.

Congratulations to @mlavacca on being promoted into the core Gateway API
maintainer team!

Thanks to @keithmattix for all your work as one of the GAMMA leads,
and congratulations to @mikemorris on being promoted into the
GAMMA lead team.

We've added two new GEP Reviewers: @kflynn and @arkodg

Also promoted in the conformance team:
* @LiorLieberman is now an Approver
* @candita is now a Reviewer

Last but most certainly not least, @guicassolato has become a reviewer for `gwctl`.

Congratulations to everyone on the promotions, and thanks for your continued
contributions to the Gateway API community!

> Relevant PRs:
> * Replace keithmattix as a mesh lead and replace with mikemorris by @keithmattix in [#3109](https://github.com/kubernetes-sigs/gateway-api/pull/3109)

Reviewer: ```suggestion
Relevant PRs:

* Replace keithmattix as a mesh lead and replace with mikemorris by @keithmattix in [#3109](https://github.com/kubernetes-sigs/gateway-api/pull/3109)
```

* fix: change report API version by @mlavacca in [#3153](https://github.com/kubernetes-sigs/gateway-api/pull/3153)
* Conformance Reporting to reflect v1.1.0 release by @xtineskim in [#3088](https://github.com/kubernetes-sigs/gateway-api/pull/3088)
* Add timeout between conformance test cases by @wstcliyu in [#3243](https://github.com/kubernetes-sigs/gateway-api/pull/3243)
* chore: added hack/verify-reports.sh script by @mlavacca in [#3167](https://github.com/kubernetes-sigs/gateway-api/pull/3167)
* feat: add useragent to conformance test run by @BobyMCbobs in [#3211](https://github.com/kubernetes-sigs/gateway-api/pull/3211)
* Expose peer certificates in round tripper response by @snehachhabria in [#3110](https://github.com/kubernetes-sigs/gateway-api/pull/3110)
* Update verify-all.sh:  added $t to explicitly tell which script had an issue during a failed test by @TheInvincibleRalph in [#3286](https://github.com/kubernetes-sigs/gateway-api/pull/3286)
* chore: golangci.yml update by @mlavacca in [#3317](https://github.com/kubernetes-sigs/gateway-api/pull/3317)

> ## New Contributors

Reviewer: 👏  We should always have this section, nicely done

### Named Route Rules

All Route Rule types (GRPCRouteRule, HTTPRouteRule, TCPRouteRule, TLSRouteRule
and UDPRouteRule) have had a new, optional `name` field to support referencing
individual rules by name.

This name, if present, may be used in `status` and logging to indicate which route
rule is being referenced in messages in a more readable way than an array index.

This feature has graduated to Experimental amd should now be used for
testing and verification purposes only. Experimental features may be
changed or removed in a future version.

This feature does not currently have a feature name defined.

This feature's status is **Extended**, meaning that it is _optional_ for
implementations to support.

As there is no feature name or conformance tests available for this feature
as yet, please see your implementation's documentation to see if it is
supported.

Relevant PRs:

* Implement of GEP-995 by @howardjohn in [#2985](https://github.com/kubernetes-sigs/gateway-api/pull/2985)
* Docs: route rule name by @guicassolato in [3299](https://github.com/kubernetes-sigs/gateway-api/pull/3299)

## Other specification changes

> * SupportedFeatures field in Gateway Class status are now presented as a list of objects by @LiorLieberman in [#3200](https://github.com/kubernetes-sigs/gateway-api/pull/3200)

Reviewer 2: This is a breaking change of an experimental API, although we called this out in the PR and multiple community meetings it may be worth highlighting this in a bigger breaking change section
